File: libs/go/area.go
package stonecharts

import (
	"fmt"
	"strconv"
	"strings"
)

// renderAreaSVG is a one-line delegation to the shared cartesian frame:
// point x-scale, include-zero value axis, a11y noun "Area".
func renderAreaSVG(spec *ChartSpec) string {
	return renderCartesian(spec, "Area", "point", areaMarks, true)
}

func areaTopPath(pts [][2]float64, step string) string {
	return pathD(pts, step)
}

func areaPath(topPts, bottomPts [][2]float64, step string) string {
	topD := areaTopPath(topPts, step)
	bottomRev := make([][2]float64, len(bottomPts))
	for i := range bottomPts {
		bottomRev[i] = bottomPts[len(bottomPts)-1-i]
	}
	bottomD := areaTopPath(bottomRev, step)
	if strings.HasPrefix(bottomD, "M") {
		bottomD = "L" + bottomD[1:]
	}
	return topD + " " + bottomD + " Z"
}

func areaSeriesFill(st seriesStyle) string {
	if st.areaFill != "" {
		return st.areaFill
	}
	if strings.HasPrefix(st.stroke, "url(") {
		return st.stroke
	}
	return st.solid
}

func areaMarks(f *cartesianFrame, p *strings.Builder) {
	if f.n <= 0 {
		return
	}

	stacked := f.stacking == "normal" || f.stacking == "percent"
	totals := make([]float64, f.n)
	if f.stacking == "percent" {
		for _, s := range f.spec.Series {
			for i, v := range s.Data {
				if i < f.n {
					totals[i] += v
				}
			}
		}
	}

	running := make([]float64, f.n)
	for si, s := range f.spec.Series {
		st := f.styles[si]
		p.WriteString(fmt.Sprintf(`<g class="sc-series" data-series="%d">`, si))

		rawVals := make([]float64, 0, len(s.Data))
		for i, v := range s.Data {
			if i >= f.n {
				break
			}
			rawVals = append(rawVals, v)
		}
		topPts := make([][2]float64, len(rawVals))

		if stacked {
			vals := make([]float64, len(rawVals))
			for i, raw := range rawVals {
				if f.stacking == "percent" {
					total := totals[i]
					if total != 0 {
						vals[i] = raw / total * 100.0
					}
				} else {
					vals[i] = raw
				}
			}
			bottomPts := make([][2]float64, len(rawVals))
			for i := range vals {
				bottomPts[i] = [2]float64{f.xpix(i), f.ypix(running[i])}
				running[i] += vals[i]
				topPts[i] = [2]float64{f.xpix(i), f.ypix(running[i])}
			}
			if len(topPts) > 0 {
				fillOp := st.areaOp
				if fillOp == "" {
					fillOp = ` fill-opacity="0.75"`
				}
				p.WriteString(fmt.Sprintf(
					`<path class="sc-series-area" data-series="%d" d="%s" fill="%s"%s stroke="none"/>`,
					si, areaPath(topPts, bottomPts, s.Step), areaSeriesFill(st), fillOp))
				lineDash := dashArray(s.DashStyle)
				lineDashAttr := ""
				if lineDash != "" {
					lineDashAttr = ` stroke-dasharray="` + lineDash + `"`
				}
				p.WriteString(fmt.Sprintf(
					`<path class="sc-series-line" data-series="%d" d="%s" fill="none" stroke="%s" stroke-width="%s" stroke-linejoin="round" stroke-linecap="round"%s/>`,
					si, areaTopPath(topPts, s.Step), st.stroke, fmtNum(s.lineWidth()), lineDashAttr))
			}
		} else {
			bottomPts := make([][2]float64, len(rawVals))
			for i, raw := range rawVals {
				topPts[i] = [2]float64{f.xpix(i), f.ypix(raw)}
				bottomPts[i] = [2]float64{f.xpix(i), f.ypix(0.0)}
			}
			if len(topPts) > 0 {
				base := f.ypix(0.0)
				areaD := areaTopPath(topPts, s.Step) + " L" + f1(topPts[len(topPts)-1][0]) + " " + f1(base) +
					" L" + f1(topPts[0][0]) + " " + f1(base) + " Z"
				fillOp := st.areaOp
				if fillOp == "" {
					fillOp = ` fill-opacity="0.75"`
				}
				p.WriteString(fmt.Sprintf(
					`<path class="sc-series-area" data-series="%d" d="%s" fill="%s"%s stroke="none"/>`,
					si, areaD, areaSeriesFill(st), fillOp))
				lineDash := dashArray(s.DashStyle)
				lineDashAttr := ""
				if lineDash != "" {
					lineDashAttr = ` stroke-dasharray="` + lineDash + `"`
				}
				p.WriteString(fmt.Sprintf(
					`<path class="sc-series-line" data-series="%d" d="%s" fill="none" stroke="%s" stroke-width="%s" stroke-linejoin="round" stroke-linecap="round"%s/>`,
					si, areaTopPath(topPts, s.Step), st.stroke, fmtNum(s.lineWidth()), lineDashAttr))
			}
		}

		mk := s.Marker
		if mk == nil || mk.Enabled == nil || *mk.Enabled {
			radius := s.markerRadius()
			radiusHover := radius + 2.5
			symbol := s.markerSymbol()
			for i, pt := range topPts {
				xlabel := strconv.Itoa(i)
				if i < len(f.cats) {
					xlabel = f.cats[i]
				}
				common := fmt.Sprintf(
					`class="sc-point" data-series="%d" data-series-name="%s" data-x="%s" data-y="%s" data-color="%s" data-r="%s" data-r-hover="%s"`,
					si, esc(s.Name), esc(xlabel), esc(fmtNum(rawVals[i])), st.solid, fmtNum(radius), fmtNum(radiusHover))
				p.WriteString(markerSVG(symbol, pt[0], pt[1], radius, common, st.solid, f.theme.MarkerHalo))
			}
		}
		p.WriteString(`</g>`)
	}
}

File: libs/python/stonecharts/charts/area.py
"""Area chart renderer: ChartSpec -> SVG string.

Shared cartesian chrome comes from _cartesian.py. This module draws the area
fill, the top-edge line, and markers for each series.
"""

from __future__ import annotations

from ..spec import Marker
from ..util import esc, fmt_num
from ._cartesian import CartesianFrame, dash_array, render_cartesian
from .line import _marker, _path_d, _spline_d


def render_svg(spec) -> str:
    return render_cartesian(spec, "Area", "point", _area_marks)


def _top_path(pts, step, curve) -> str:
    return _spline_d(pts) if curve == "monotone" else _path_d(pts, step)


def _area_path(top_pts, bottom_pts, step, curve) -> str:
    top_d = _top_path(top_pts, step, curve)
    bottom_d = _top_path(list(reversed(bottom_pts)), step, curve)
    if bottom_d.startswith("M"):
        bottom_d = "L" + bottom_d[1:]
    return f"{top_d} {bottom_d} Z"


def _series_fill(st) -> str:
    if st.area_fill is not None:
        return st.area_fill
    if st.stroke.startswith("url("):
        return st.stroke
    return st.solid


def _area_marks(fr: CartesianFrame, p: list[str]) -> None:
    if fr.n <= 0:
        return

    stacked = fr.stacking in ("normal", "percent")
    totals = [0.0] * fr.n
    if fr.stacking == "percent":
        for s in fr.spec.series:
            for i, v in enumerate(s.data):
                if i < fr.n:
                    totals[i] += v

    running = [0.0] * fr.n
    theme = fr.theme
    for si, s in enumerate(fr.spec.series):
        st = fr.styles[si]
        raw_vals = [float(v) for v in s.data[: fr.n]]
        top_pts = []
        p.append(f'<g class="sc-series" data-series="{si}">')

        if stacked:
            vals = []
            for i, raw in enumerate(raw_vals):
                if fr.stacking == "percent":
                    total = totals[i]
                    vals.append(0.0 if total == 0.0 else raw / total * 100.0)
                else:
                    vals.append(raw)
            bottom_vals = running[: len(vals)]
            top_vals = [bottom_vals[i] + vals[i] for i in range(len(vals))]
            running[: len(vals)] = top_vals
            top_pts = [(fr.xpix(i), fr.ypix(top_vals[i])) for i in range(len(vals))]
            bottom_pts = [(fr.xpix(i), fr.ypix(bottom_vals[i])) for i in range(len(vals))]
            if top_pts:
                fill_op = st.area_op if st.area_op else ' fill-opacity="0.75"'
                p.append(
                    f'<path class="sc-series-area" data-series="{si}" d="{_area_path(top_pts, bottom_pts, s.step, s.curve)}" '
                    f'fill="{_series_fill(st)}"{fill_op} stroke="none"/>'
                )
                line_dash = dash_array(s.dash_style)
                line_dash_attr = f' stroke-dasharray="{line_dash}"' if line_dash else ""
                p.append(
                    f'<path class="sc-series-line" data-series="{si}" d="{_top_path(top_pts, s.step, s.curve)}" fill="none" '
                    f'stroke="{st.stroke}" stroke-width="{fmt_num(s.line_width if s.line_width is not None else 2)}" '
                    f'stroke-linejoin="round" stroke-linecap="round"{line_dash_attr}/>'
                )
        else:
            top_pts = [(fr.xpix(i), fr.ypix(raw_vals[i])) for i in range(len(raw_vals))]
            if top_pts:
                base = fr.ypix(0.0)
                area_d = f"{_top_path(top_pts, s.step, s.curve)} L{top_pts[-1][0]:.1f} {base:.1f} L{top_pts[0][0]:.1f} {base:.1f} Z"
                fill_op = st.area_op if st.area_op else ' fill-opacity="0.75"'
                p.append(
                    f'<path class="sc-series-area" data-series="{si}" d="{area_d}" '
                    f'fill="{_series_fill(st)}"{fill_op} stroke="none"/>'
                )
                line_dash = dash_array(s.dash_style)
                line_dash_attr = f' stroke-dasharray="{line_dash}"' if line_dash else ""
                p.append(
                    f'<path class="sc-series-line" data-series="{si}" d="{_top_path(top_pts, s.step, s.curve)}" fill="none" '
                    f'stroke="{st.stroke}" stroke-width="{fmt_num(s.line_width if s.line_width is not None else 2)}" '
                    f'stroke-linejoin="round" stroke-linecap="round"{line_dash_attr}/>'
                )

        mk = s.marker or Marker()
        if mk.enabled and top_pts:
            radius = mk.radius
            radius_hover = radius + 2.5
            for i, (x, y) in enumerate(top_pts):
                xlabel = fr.cats[i] if i < len(fr.cats) else str(i)
                common = (
                    f'class="sc-point" data-series="{si}" data-series-name="{esc(s.name)}" '
                    f'data-x="{esc(xlabel)}" data-y="{esc(fmt_num(raw_vals[i]))}" '
                    f'data-color="{st.solid}" data-r="{fmt_num(radius)}" '
                    f'data-r-hover="{fmt_num(radius_hover)}"'
                )
                p.append(_marker(mk.symbol, x, y, radius, common, st.solid, theme.marker_halo))
        p.append("</g>")
